pkg/boot/linux.go

```diff
@@ -48,10 +49,79 @@ type LoadedLinuxImage struct {
 	KexecOpts   linux.KexecOptions
 }
 
+// loadedLinuxImageJSON is same as LoadedLinuxImage, but with transformed fields to help with serialization of LoadedLinuxImage.
+type loadedLinuxImageJSON struct {
+	Name        string
+	KernelPath  string
+	InitrdPath  string
+	Cmdline     string
+	LoadSyscall bool
+	KexecOpts   linux.KexecOptions
+}
+
 var _ OSImage = &LinuxImage{}
 
 var errNilKernel = errors.New("kernel image is empty, nothing to execute")
 
+// MarshalJSON customizes marshaling for LoadedLinuxImage. It handles serializations
+// for *os.File, so that kernel and initrd can be unmarshalled properly.
+func (lli *LoadedLinuxImage) MarshalJSON() ([]byte, error) {
+	lliJSON := loadedLinuxImageJSON{}
+	// Sync and close kernel and initrd File object, and marshal paths to the files.
+	if lli.Kernel != nil {
```

Comment:
why are these closing these objects? the lifetime should be managed by the caller

Reply:
Here is my view:  

1) the premise of this serialization is that we are not kexec loading, and exit early. So there is no further logic from caller.
2) we want to make sure all contents there are are captured on stable storage, so future unmarshall accurately reflects current os.File -- so I'd prefer we render the file descriptor un-usable immediately. 
3) it is good to do 2) b/c of 1)

Reply:
I don't really agree, because this makes it hard to figure out what boot/pxeboot commands should do. They are now going to have to do different things based on whether Load did this json marshal logic or not.

boot/pxeboot (callers in general) should be responsible for closing all files in all situations here

Reply:
I don't follow how harder it makes boot/pxeboot commands -- elaborate ? / examples ?

Reply:
nevermind actually, i'm good with this. lifetime is constrainted to the Load function, so this seems good.